client/src/portal/api/legacy_keyring.rs

```diff
@@ -0,0 +1,272 @@
+//! Legacy GNOME Keyring file format low level API.
+
+#[cfg(feature = "async-std")]
+use std::io;
+use std::{
+    collections::HashMap,
+    io::{Cursor, Read},
+};
+
+use endi::{Endian, ReadBytes};
+#[cfg(feature = "tokio")]
+use tokio::io;
+
+use super::{Item, Secret};
+use crate::{
+    crypto,
+    portal::{AttributeValue, Error, WeakKeyError},
+    AsAttributes,
+};
+
+const FILE_HEADER: &[u8] = b"GnomeKeyring\n\r\0\n";
+const FILE_HEADER_LEN: usize = FILE_HEADER.len();
+
+pub const MAJOR_VERSION: u8 = 0;
+pub const MINOR_VERSION: u8 = 0;
+
+#[derive(Debug)]
+pub struct Keyring {
+    salt: Vec<u8>,
+    iteration_count: u32,
+    encrypted_content: Vec<u8>,
+    item_count: usize,
+}
+
+impl Keyring {
+    pub fn decrypt_items(self, secret: &Secret) -> Result<Vec<Item>, Error> {
+        let (key, iv) = crypto::legacy_derive_key_and_iv(
+            &**secret,
+            self.key_strength(secret),
+            &self.salt,
+            self.iteration_count.try_into().unwrap(),
+        );
+        let decrypted = crypto::decrypt_no_padding(&self.encrypted_content, &key, iv);
+        let (digest, content) = decrypted.split_at(16);
+        if !crypto::verify_checksum_md5(digest, content) {
+            // This not a MAC but anyway
```

Comment:
Would be nice to have a specific error for this instead of using the wrong one, would make it easier to track down any potential issues in the future

Reply:
the other option would be to rename the used error to something more generic and maybe make it take a `&'static str` for the algorithm

Reply:
Added `ChecksumMismatch` error.